File: src/org/genyris/java/swing/GenyrisTableModel.java
package org.genyris.java.swing;

import javax.swing.table.AbstractTableModel;

import org.genyris.core.Bignum;
import org.genyris.core.Exp;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.AbstractClosure;
import org.genyris.interp.Environment;
import org.genyris.interp.StandardEnvironment;
import org.genyris.java.JavaUtils;

public class GenyrisTableModel extends AbstractTableModel {

	private static final long serialVersionUID = 2222008167725577587L;
	private AbstractClosure callback;

	public GenyrisTableModel(AbstractClosure callback) {
		this.callback = callback;
	}

	public int getColumnCount() {
		return callGenyrisInt("getColumnCount");
	}

	private int callGenyrisInt(String symbolName) {
		try {
			Exp[] arguments = {};
			Exp closure = runtimeEnv().lookupVariableValue(runtimeEnv().internString(symbolName));
			Exp result = closure.applyFunction(runtimeEnv(), arguments);
			return ((Bignum)result).bigDecimalValue().intValue();
		} catch (GenyrisException e) {
			e.printStackTrace();
		}
		return 0;
	}

	private Environment runtimeEnv() {
		return new StandardEnvironment(callback.getEnv());
	}

	public int getRowCount() {
		return callGenyrisInt("getRowCount");
	}

	public Object getValueAt(int arg0, int arg1) {
		try {
			Exp[] arguments = { new Bignum(arg0), new Bignum(arg1)};
			Exp closure = runtimeEnv().lookupVariableValue(runtimeEnv().internString("getValueAt"));
			Exp result = closure.applyFunction(runtimeEnv(), arguments);
			return JavaUtils.convertToJava(Object.class, result, runtimeEnv());
		} catch (GenyrisException e) {
			e.printStackTrace();
		}
		return null;
	}

	public void setValueAt(Object aValue, int rowIndex, int columnIndex) {
		try {
			Exp[] arguments = { new Bignum(rowIndex), 
					new Bignum(columnIndex), 
					JavaUtils.javaToGenyris(runtimeEnv(), aValue)};
			Exp closure = runtimeEnv().lookupVariableValue(runtimeEnv().internString("setValueAt"));
			closure.applyFunction(runtimeEnv(), arguments);
		} catch (GenyrisException e) {
			e.printStackTrace();
		}

	}

	public boolean isCellEditable(int rowIndex, int columnIndex) {
		return true;
	}

	public String getColumnName(int column) {
		return Integer.toString(column);
	}
}

File: src/org/genyris/java/swing/GenyrisTableModel.kt
package org.genyris.java.swing

import org.genyris.core.Bignum
import org.genyris.core.Exp
import org.genyris.exception.GenyrisException
import org.genyris.interp.AbstractClosure
import org.genyris.interp.Environment
import org.genyris.interp.StandardEnvironment
import org.genyris.java.JavaUtils
import javax.swing.table.AbstractTableModel

class GenyrisTableModel(private val callback: AbstractClosure) : AbstractTableModel() {
    override fun getColumnCount(): Int {
        return callGenyrisInt("getColumnCount")
    }

    private fun callGenyrisInt(symbolName: String?): Int {
        try {
            val arguments = arrayOf<Exp?>()
            val closure = runtimeEnv().lookupVariableValue(runtimeEnv().internString(symbolName))
            val result = closure.applyFunction(runtimeEnv(), arguments)
            return (result as Bignum).bigDecimalValue().intValue()
        } catch (e: GenyrisException) {
            e.printStackTrace()
        }
        return 0
    }

    private fun runtimeEnv(): Environment {
        return StandardEnvironment(callback.getEnv())
    }

    override fun getRowCount(): Int {
        return callGenyrisInt("getRowCount")
    }

    override fun getValueAt(arg0: Int, arg1: Int): Any? {
        try {
            val arguments = arrayOf<Exp?>(Bignum(arg0), Bignum(arg1))
            val closure = runtimeEnv().lookupVariableValue(runtimeEnv().internString("getValueAt"))
            val result = closure.applyFunction(runtimeEnv(), arguments)
            return JavaUtils.convertToJava(Any::class.java, result, runtimeEnv())
        } catch (e: GenyrisException) {
            e.printStackTrace()
        }
        return null
    }

    override fun setValueAt(aValue: Any?, rowIndex: Int, columnIndex: Int) {
        try {
            val arguments = arrayOf<Exp?>(
                Bignum(rowIndex),
                Bignum(columnIndex),
                JavaUtils.javaToGenyris(runtimeEnv(), aValue)
            )
            val closure = runtimeEnv().lookupVariableValue(runtimeEnv().internString("setValueAt"))
            closure.applyFunction(runtimeEnv(), arguments)
        } catch (e: GenyrisException) {
            e.printStackTrace()
        }
    }

    override fun isCellEditable(rowIndex: Int, columnIndex: Int): Boolean {
        return true
    }

    override fun getColumnName(column: Int): String {
        return Integer.toString(column)
    }

    companion object {
        private const val serialVersionUID = 2222008167725577587L
    }
}
